File: backend/src/main/java/com/impaintor/feature/realtime/dto/outbound/GameEvent.java
package com.impaintor.feature.realtime.dto.outbound;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

/**
 * Evento del juego difundido a {@code /topic/room.{code}.game}.
 *
 * <p>Discriminador {@code type} (ver sección 6.2 de CLAUDE.md). Los Tracks H/I
 * conocerán solo {@link GameEvent} — no necesitan tocar STOMP/Jackson.</p>
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = GameEvent.GameStart.class, name = "GAME_START"),
        @JsonSubTypes.Type(value = GameEvent.TurnStart.class, name = "TURN_START"),
        @JsonSubTypes.Type(value = GameEvent.TurnEnd.class, name = "TURN_END"),
        @JsonSubTypes.Type(value = GameEvent.GalleryPhase.class, name = "GALLERY_PHASE"),
        @JsonSubTypes.Type(value = GameEvent.VotePhase.class, name = "VOTE_PHASE"),
        @JsonSubTypes.Type(value = GameEvent.VoteResult.class, name = "VOTE_RESULT"),
        @JsonSubTypes.Type(value = GameEvent.VoteTie.class, name = "VOTE_TIE"),
        @JsonSubTypes.Type(value = GameEvent.GuessAttempt.class, name = "GUESS_ATTEMPT"),
        @JsonSubTypes.Type(value = GameEvent.NewRound.class, name = "NEW_ROUND"),
        @JsonSubTypes.Type(value = GameEvent.GameOver.class, name = "GAME_OVER")
})
public sealed interface GameEvent {

    record TopVote(Long id, int votes) {}

    record GameStart(List<Long> drawingOrder, int round) implements GameEvent {}

    record TurnStart(Long playerId, int timeSeconds) implements GameEvent {}

    record TurnEnd(Long playerId) implements GameEvent {}

    record GalleryPhase() implements GameEvent {}

    record VotePhase(int timeSeconds) implements GameEvent {}

    record VoteResult(Long eliminated, boolean wasImpostor, List<TopVote> topVoted) implements GameEvent {}

    record VoteTie(List<TopVote> tiedPlayers, int timeSeconds) implements GameEvent {}

    record GuessAttempt(int livesRemaining, boolean correct) implements GameEvent {}

    record NewRound(int round, List<Long> drawingOrder) implements GameEvent {}

    record GameOver(String winner, String reason, Long impostorId, String secretWord) implements GameEvent {}
}

File: frontend/src/app/features/game/models/game-event.ts
// Espejo TypeScript del sealed interface GameEvent del backend Track D.
// Mantener sincronizado con:
//   backend/src/main/java/com/impaintor/feature/realtime/dto/outbound/GameEvent.java

export interface TopVote {
  id: number;
  votes: number;
}

export type EndReason =
  | 'VOTED_OUT'
  | 'WORD_GUESSED'
  | 'OUT_OF_LIVES'
  | 'TIE_NOT_BROKEN'
  | 'LAST_STANDING';

export type Winner = 'PAINTERS' | 'IMPOSTOR';

/** Discriminated union de todos los eventos del juego que el servidor difunde a /topic/room.{code}.game. */
export type GameEvent =
  | { type: 'GAME_START'; drawingOrder: number[]; round: number }
  | { type: 'TURN_START'; playerId: number; timeSeconds: number; drawingOrder: number[] }
  | { type: 'TURN_END'; playerId: number }
  | { type: 'GALLERY_PHASE' }
  | { type: 'VOTE_PHASE'; timeSeconds: number }
  | { type: 'VOTE_RESULT'; eliminated: number | null; wasImpostor: boolean; topVoted: TopVote[] }
  | { type: 'VOTE_TIE'; tiedPlayers: TopVote[]; timeSeconds: number }
  | { type: 'GUESS_ATTEMPT'; livesRemaining: number; correct: boolean }
  | { type: 'NEW_ROUND'; round: number; drawingOrder: number[] }
  | { type: 'GAME_OVER'; winner: Winner; reason: EndReason; impostorId: number; secretWord: string; eloChange?: number };

/** Strokes y limpiezas que el servidor reenvía a /topic/room.{code}.draw. */
export interface StrokeBroadcast {
  type: 'STROKE';
  playerId: number;
  points: { x: number; y: number }[];
  color: string;
  thickness: number;
}

export interface ClearCanvasBroadcast {
  type: 'CLEAR';
  playerId: number;
}

export type DrawBroadcast = StrokeBroadcast | ClearCanvasBroadcast;

/**
 * Comandos de dibujo que el cliente envía a /app/room.{code}.draw.
 * El servidor añade el `playerId` desde el Principal autenticado (anti-spoofing),
 * por eso aquí no aparece.
 */
export type DrawCommand =
  | { type: 'STROKE'; points: { x: number; y: number }[]; color: string; thickness: number }
  | { type: 'CLEAR' };
